tests/unit/instance-initializers/prefetch-test.js

```diff
@@ -48,3 +49,79 @@ test('the prefetch hook is invoked', function(assert) {
   assert.equal(handler1._prefetched, promise, 'the promise is set on handler1 as _prefetched');
   assert.equal(handler2._prefetched, promise, 'the promise is set on handler2 as _prefetched');
 });
+
+test('the prefetch hook is invoked for async handlers', function(assert) {
+  assert.expect(4);
+
+  initialize(instance);
+
+  const router = instance.lookup('router:main');
+  const promise = new Ember.RSVP.resolve(1);
+  const runSharedModelHook = this.stub().returns(promise);
+  const handler1 = {};
+  const handler2 = {};
+  const handlerInfos = [{
+    runSharedModelHook,
+    handler: handler1,
+  }, {
+    runSharedModelHook,
+    handler: undefined,
+    handlerPromise: new Ember.RSVP.resolve(handler2)
+  }];
+  const transition = { handlerInfos };
+
+  router.trigger('willTransition', transition);
+
+  return wait().then(() => {
+    assert.ok(runSharedModelHook.calledTwice, 'handlerInfo.runSharedModelHook is called once per handlerInfo');
+    assert.ok(runSharedModelHook.calledWith(transition, 'prefetch'), 'handlerInfo.runSharedModelHook is called with the transition and "prefetch"');
+    assert.equal(handler1._prefetched, promise, 'the promise is set on handler1 as _prefetched');
+    assert.equal(handler2._prefetched, promise, 'the promise is set on handler2 as _prefetched');
+  });
+});
+
+test('the prefetch hook is invoked according to route hierarchy for async handlers', function(assert) {
+  assert.expect(1);
+
+  initialize(instance);
+
+  const router = instance.lookup('router:main');
+  const operations = [];
+  const handler1 = {};
+  const handler2 = {};
+  const handlerInfos = [{
+    runSharedModelHook() {
+      operations.push('prefetch 1');
+    },
+    handler: undefined,
+    handlerPromise: new Ember.RSVP.Promise((resolve) => {
+      setTimeout(() => {
+        operations.push('resolved handler 1');
+        resolve(handler1);
+      }, 50);
+    })
+  }, {
+    runSharedModelHook() {
+      operations.push('prefetch 2');
+    },
+    handler: undefined,
+    handlerPromise: new Ember.RSVP.Promise((resolve) => {
+      setTimeout(() => {
+        operations.push('resolved handler 2');
+        resolve(handler2);
+      }, 0);
+    })
+  }];
+  const transition = { handlerInfos };
+
+  router.trigger('willTransition', transition);
+
+  return transition._handlerPromiseChain.then(() => {
```

Comment:
You have access to the handler promises here (because you're creating them here). Let's avoid accessing the handler promise chain from the transition, which in turn avoids the need to store it on the transition.

This will require defining the promises outside the handler array/object literals, but has beneficial side effects. They can be defined in reverse order, 2 before 1, and therefore can both use a timeout with delay 0 (since ordering on the MacroTask queue will still be 2 before 1), so we can save those 50ms.